compiler/rustc_typeck/src/check/check.rs

```diff
@@ -1590,3 +1591,31 @@ fn opaque_type_cycle_error(tcx: TyCtxt<'tcx>, def_id: LocalDefId, span: Span) {
     }
     err.emit();
 }
+
+struct SelfTySpanVisitor<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    selfty_spans: Vec<Span>,
+}
+
+impl Visitor<'tcx> for SelfTySpanVisitor<'tcx> {
+    type Map = rustc_middle::hir::map::Map<'tcx>;
+
+    fn nested_visit_map(&mut self) -> hir::intravisit::NestedVisitorMap<Self::Map> {
+        hir::intravisit::NestedVisitorMap::OnlyBodies(self.tcx.hir())
+    }
+
+    fn visit_ty(&mut self, arg: &'tcx hir::Ty<'tcx>) {
+        match arg.kind {
+            hir::TyKind::Path(hir::QPath::Resolved(None, path)) => match &path.segments {
+                [segment]
+                    if segment.res.map(|res| matches!(res, Res::SelfTy(_, _))).unwrap_or(false) =>
```

Comment:
nit: I think you could write this as

```suggestion
                [PathSegment { res: Res::SelfTy(_, _), .. }] => {
```